q2_quality_control/_threshold_graph/_visualizer.py

```diff
@@ -60,30 +76,125 @@ def decontam_score_viz(output_dir, decontam_scores: qiime2.Metadata,
     gray_lab_arr = []
     red_lab_arr = []
     blue_lab_arr = []
+    data_arr = []
+    display_seqs_arr = []
+    true_fasta_dest = []
+    contam_fasta_dest = []
+    nan_fasta_dest = []
+    sorted_key_arr = []
+    feature_or_read_arr = []
 
     for key in table_dict.keys():
         table = table_dict[key]
         decontam_scores = decontam_scores_dict[key]
-        df = decontam_scores.to_dataframe()
+        df = decontam_scores
         if df['p'].isna().all():
             raise ValueError("No p-values exist for the data provided.")
 
         read_nums = table.sum(axis='rows')
-
         p_vals = df['p'].dropna()
         filt_read_nums = read_nums[p_vals.index]
 
         contams = (p_vals < threshold)
+        contam_indices = []
```

Comment:
This section would benefit from some comments, and I think a lot of this could be refactored to pandas commands (e.g., you can slice based on boolean arrays, which seems to be what you're doing here). Can you take another pass through this and try to simplify and clarify it?